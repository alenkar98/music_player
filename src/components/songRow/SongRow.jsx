import { FaCaretDown, FaHeart, FaPause, FaPlay } from 'react-icons/fa';
import { TiTick } from 'react-icons/ti';
import { LuForward } from 'react-icons/lu';
import './songRow.css';
import { useEffect, useRef, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { playNextSong } from '../../redux/action/songAction';

export default function SongRow({ name, artist, songIndex, songLink, onQueue }) {
  const dispatch = useDispatch();
  const isAllSongsPlaying = useSelector((state) => state.song.isAllSongsPlaying);
  const [isPlaying, setIsPlaying] = useState(false);
  const audioRef = useRef(null);

  useEffect(() => {
    if (isAllSongsPlaying) {
      handlePlayButtonClick();
    } else {
      setIsPlaying(false);
      audioRef.current.pause();
    }
  }, [isAllSongsPlaying]);

  useEffect(() => {
    if (onQueue) {
      setIsPlaying(true);
      audioRef.current.play();
    }
  }, [onQueue]);

  const handlePlayButtonClick = () => {
    setIsPlaying(!isPlaying);
    if (audioRef.current) {
      const audioElement = audioRef.current;
      if (!isPlaying) {
        audioElement.play();
      } else {
        audioElement.pause();
      }
    }
  };

  const handleSongEnded = () => {
    setIsPlaying(false);
    dispatch(playNextSong());
  };

  return (
    <>
      <div className="row">
        <div className="list">
          <button onClick={handlePlayButtonClick}>{isPlaying ? <FaPause /> : <FaPlay />}</button>
        </div>
        <div className="songName">
          <span>{name}</span>
        </div>
        <div className="artistName">
          <span>{artist}</span>
        </div>
        <div className="track">
          <span>{songIndex}</span>
        </div>
        <div className="buttons">
          <button>
            <FaHeart />
          </button>
          <button>
            <TiTick />
          </button>
          <button>
            <LuForward />
          </button>
          <button>
            <FaCaretDown />
          </button>
        </div>
      </div>
      <div>
        <audio ref={audioRef} onEnded={handleSongEnded}>
          <source src={songLink} />
        </audio>
      </div>
    </>
  );
}
