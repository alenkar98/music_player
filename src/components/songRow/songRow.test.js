import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import SongRow from './SongRow';

const mockStore = configureStore([]);

test('SongRow component renders correctly', () => {
  const initialState = {
    song: {
      isAllSongsPlaying: false,
    },
  };

  const store = mockStore(initialState);

  const songProps = {
    name: 'Test Song',
    artist: 'Test Artist',
    songIndex: 1,
    songLink: 'test-song-link.mp3',
    onQueue: false,
  };

  const component = renderer.create(
    <Provider store={store}>
      <SongRow {...songProps} />
    </Provider>,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
