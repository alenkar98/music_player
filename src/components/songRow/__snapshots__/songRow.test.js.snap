// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongRow component renders correctly 1`] = `
Array [
  <div
    className="row"
  >
    <div
      className="list"
    >
      <button
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
          />
        </svg>
      </button>
    </div>
    <div
      className="songName"
    >
      <span>
        Test Song
      </span>
    </div>
    <div
      className="artistName"
    >
      <span>
        Test Artist
      </span>
    </div>
    <div
      className="track"
    >
      <span>
        1
      </span>
    </div>
    <div
      className="buttons"
    >
      <button>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
          />
        </svg>
      </button>
      <button>
        <svg
          baseProfile="tiny"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          version="1.2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16.972 6.251c-.967-.538-2.185-.188-2.72.777l-3.713 6.682-2.125-2.125c-.781-.781-2.047-.781-2.828 0-.781.781-.781 2.047 0 2.828l4 4c.378.379.888.587 1.414.587l.277-.02c.621-.087 1.166-.46 1.471-1.009l5-9c.537-.966.189-2.183-.776-2.72z"
          />
        </svg>
      </button>
      <button>
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="15 17 20 12 15 7"
          />
          <path
            d="M4 18v-2a4 4 0 0 1 4-4h12"
          />
        </svg>
      </button>
      <button>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 320 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          />
        </svg>
      </button>
    </div>
  </div>,
  <div>
    <audio
      onEnded={[Function]}
    >
      <source
        src="test-song-link.mp3"
      />
    </audio>
  </div>,
]
`;
